app/src/test/java/org/mozilla/fenix/components/metrics/AppStartupTelemetryTest.kt

```diff
@@ -0,0 +1,163 @@
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Source.APP_ICON
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Source.UNKNOWN
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type.COLD
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type.WARM
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type.HOT
+
+class AppStartupTelemetryTest {
+
+    @RelaxedMockK
+    private lateinit var metricController: MetricController
+    @RelaxedMockK
+    private lateinit var intent: Intent
+
+    private lateinit var appStartupTelemetry: AppStartupTelemetry
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this)
+        appStartupTelemetry = AppStartupTelemetry(metricController)
+    }
+
+    @Test
+    fun `WHEN application is launch for the first time through application icon THEN startup metric records {APP_ICON, COLD, FALSE}`() {
+        every { intent.action } returns Intent.ACTION_MAIN
+        every { intent.categories.contains(Intent.CATEGORY_LAUNCHER) } returns true
+
+        appStartupTelemetry.onFenixApplicationOnCreate()
+        appStartupTelemetry.onHomeActivityOnCreate(intent.toSafeIntent(), false)
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppStartupSourceAndType(APP_ICON, COLD, false)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `WHEN application is launch for the first time through a url link THEN startup metric records {LINK, COLD, FALSE}`() {
+        every { intent.action } returns Intent.ACTION_VIEW
+
+        appStartupTelemetry.onFenixApplicationOnCreate()
+        appStartupTelemetry.onHomeActivityOnCreate(intent.toSafeIntent(), false)
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppStartupSourceAndType(LINK, COLD, false)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `WHEN application is launch for the first time through an external app THEN startup metric records {CUSTOM_TAB, COLD, FALSE}`() {
+        appStartupTelemetry.onFenixApplicationOnCreate()
+        appStartupTelemetry.onExternalAppBrowserOnCreate(intent.toSafeIntent(), false)
+        appStartupTelemetry.onHomeActivityOnResume()
+
+        val validMetric = AppStartupSourceAndType(CUSTOM_TAB, COLD, false)
+        verify(exactly = 1) { metricController.track(validMetric) }
+    }
+
+    @Test
+    fun `GIVEN, application  exists and is backgrounded, WHEN application is launched again through app icon and HomeActivity is recreated THEN startup metric records {APP_ICON, WARM, FALSE}`() {
+        every { intent.action } returns Intent.ACTION_MAIN
+        every { intent.categories.contains(Intent.CATEGORY_LAUNCHER) } returns true
+        appStartupTelemetry.onApplicationOnStop()
```

Comment:
I think you need to do `onCreate; onCreate; onResume; onStop` (or at least `onCreate; onStop` if HomeActivity isn't launched) to get the test set up correctly, right? Otherwise, can we be sure about the state of the object assuming we don't know about the implementation?

This "application is backgrounded" is also a good candidate for a helper function.

Reply:
done.